kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -551,6 +559,51 @@ noConfusionInjectionConstructor
 noConfusionInjectionConstructor =
     explainAndReturnBottom "No confusion: sort injections and constructors"
 
+{- |
+ If the two constructors form an overload pair, apply the overloading axioms
+ on the terms to make the constructors equal, then retry unification on them.
+
+See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
+
+ -}
+overloadedConstructorSortInjectionAndEquals
+    :: (InternalVariable variable, MonadUnify unifier)
+    => TermSimplifier variable unifier
+    -> TermLike variable
+    -> TermLike variable
+    -> MaybeT unifier (Pattern variable)
+overloadedConstructorSortInjectionAndEquals termMerger firstTerm secondTerm
+  = do
+    eunifier <- lift . Error.runExceptT
+        $ unifyOverloading (Pair firstTerm secondTerm)
+    case eunifier of
+        Right (Simple (Pair firstTerm' secondTerm')) -> lift $
+            termMerger firstTerm' secondTerm'
+        Right
+            (WithNarrowing Narrowing
+                { narrowedVar
+                , narrowingTerm
+                , overloadPair = Pair firstTerm' secondTerm'
+                }
+            ) -> lift $ do
+                let narrowingVars =
+                        getFreeElementVariables $ freeVariables narrowingTerm
+                let narrowedSubst =
+                        Predicate.markSimplified
+                        $ Predicate.makeEqualsPredicate_
+                            (mkElemVar narrowedVar)
+                            narrowingTerm
+                merged <- termMerger firstTerm' secondTerm'
+                boundPattern <-
+                    Exists.makeEvaluate SideCondition.topTODO narrowingVars
+                        $ merged `Pattern.andCondition` from narrowedSubst
+                case OrPattern.toPatterns boundPattern of
+                    [result] -> return result
+                    _ -> empty
```

Comment:
This returns `\bottom` if there are multiple results. Is that right?

Reply:
perhaps it should error in that case. thanks for catching it!

Reply:
we changed the code such that it returns bottom for the empty list and not applicable if there is more than a result.